### ISSUE TITLE
valid shift with const value but signed type as the right operand is rejected

### ISSUE THREAD

**Issue by apinski-cavium** [p4lang/p4c#3287]
Take:
```
bit<4> func(in bit<4> l)
{
  const int<6> tt = 1;
  return l << tt;
}
```
The reference compiler rejects this but the specification reads as:
```
must be either an
  unsigned number of type `bit<S>` or a non-negative integer compile-time known value.
```
tt here is a compile time know value of 1.

Note the reference compiler also points to in the error message to the wrong location too:
```
shift.p4(3): [--Werror=expected] error: 6s1: shift amounts cannot be signed
  const int<6> tt = 1;
                    ^
```
I don't see a shift there. (that is most likely similar to https://github.com/p4lang/p4c/issues/3286).

**Comment by mbudiu-vmw** [comment]
We actually already have an explicit test issue2332-1.p4 for this.
So it looks like the spec is slightly self-contradictory or fuzzy.
We forbid shifts with signed values - period.
The question is whether we can make an exception if we know for sure that the value is constant and thus positive.
I will file an issue with the spec.

**Comment by mbudiu-vmw** [comment]
Marked as enhancement to at least give a better error message.

**Comment by mbudiu-vmw** [comment]
I have decided to allow this behavior, the spec seems to actually be fine.

**Comment by mbudiu-vmw** [comment]
Fixed by #3303
